src/librustc/infer/const_variable.rs

```diff
@@ -0,0 +1,271 @@
+use crate::mir::interpret::ConstValue;
+use syntax::symbol::InternedString;
+use syntax_pos::Span;
+use crate::ty::{self, InferConst};
+
+use std::cmp;
+use std::marker::PhantomData;
+use rustc_data_structures::fx::FxHashMap;
+use rustc_data_structures::snapshot_vec as sv;
+use rustc_data_structures::unify as ut;
+
+pub struct ConstVariableTable<'tcx> {
+    values: sv::SnapshotVec<Delegate<'tcx>>,
+
+    relations: ut::UnificationTable<ut::InPlace<ty::ConstVid<'tcx>>>,
+}
+
+/// Reasons to create a const inference variable
+#[derive(Copy, Clone, Debug)]
+pub enum ConstVariableOrigin {
+    MiscVariable(Span),
+    ConstInference(Span),
+    ConstParameterDefinition(Span, InternedString),
+    SubstitutionPlaceholder(Span),
+}
+
+pub type ConstVariableMap<'tcx> = FxHashMap<ty::ConstVid<'tcx>, ConstVariableOrigin>;
+
+struct ConstVariableData {
+    origin: ConstVariableOrigin,
+}
+
+#[derive(Copy, Clone, Debug)]
+pub enum ConstVariableValue<'tcx> {
+    Known { value: &'tcx ty::LazyConst<'tcx> },
+    Unknown { universe: ty::UniverseIndex },
+}
```

Comment:
@nikomatsakis I'm thinking maybe there should be a generic `enum` like this to share between types and consts.

Reply:
Yeah, so, I rewrote ena for lark and I made a bunch of changes. As I recall, one of the things I changed was eliminating this sort of boilerplate (I had some other goals too, like being able to track the tree of unifications for better error messages). To be honest, I sort of *forget* what I did in Lark. I think this is another work item that is worth trying to postpone till after the PR lands, and then cleanup separately.